contracts/contracts/strategies/ConvexEthMetaStrategy.sol

```diff
@@ -0,0 +1,381 @@
+         * created is no longer valid.
+         */
+
+        uint256 poolWETHBalance = curvePool.balances(wethCoinIndex);
+        /* K is multiplied by 1e36 which is used for higher precision calculation of required
+         * pool LP tokens. Without it the end value can have rounding errors up to precision of
+         * 10 digits. This way we move the decimal point by 36 places when doing the calculation
+         * and again by 36 places when we are done with it.
+         */
+        uint256 k = (1e36 * lpToken.totalSupply()) / poolWETHBalance;
+        // simplifying below to: `uint256 diff = (_wethAmount - 1) * k` causes loss of precision
+        // prettier-ignore
+        // slither-disable-next-line divide-before-multiply
+        uint256 diff = poolWETHBalance * k -
+            (poolWETHBalance - _wethAmount - 1) * k;
+        lpToBurn = diff / 1e36;
+    }
+
+    /**
+     * @dev Remove all assets from platform and send them to Vault contract.
+     */
+    function withdrawAll() external override onlyVaultOrGovernor nonReentrant {
+        _lpWithdrawAll();
+
+        // Withdraws are proportional to assets held by 3Pool
+        uint256[2] memory minWithdrawAmounts = [uint256(0), uint256(0)];
+
+        // Remove liquidity
+        curvePool.remove_liquidity(
+            lpToken.balanceOf(address(this)),
+            minWithdrawAmounts
+        );
+
+        // Burn OETH
+        IVault(vaultAddress).burnForStrategy(
+            poolOETHToken.balanceOf(address(this))
+        );
+        // Transfer assets to the Vault
+        poolWETHToken.safeTransfer(
+            vaultAddress,
+            poolWETHToken.balanceOf(address(this))
+        );
+    }
+
+    function _lpWithdraw(uint256 _wethAmount) internal {
+        // withdraw and unwrap with claim takes back the lpTokens and also collects the rewards for deposit
+        IRewardStaking(cvxRewardStakerAddress).withdrawAndUnwrap(
+            _wethAmount,
+            true
+        );
+    }
+
+    function _lpWithdrawAll() internal {
+        uint256 gaugeTokens = IRewardStaking(cvxRewardStakerAddress).balanceOf(
+            address(this)
+        );
+        _lpWithdraw(gaugeTokens);
+    }
+
+    /**
+     * @dev Get the total asset value held in the platform
+     * @param _asset      Address of the asset
+     * @return balance    Total value of the asset in the platform
+     */
+    function checkBalance(address _asset)
+        public
+        view
+        virtual
+        override
+        returns (uint256 balance)
+    {
+        require(_asset == address(poolWETHToken), "Unsupported asset");
+        // LP tokens in this contract. This should generally be nothing as we
+        // should always stake the full balance in the Gauge, but include for
+        // safety
+        uint256 totalPTokens = lpToken.balanceOf(address(this));
+        if (totalPTokens > 0) {
+            uint256 virtual_price = curvePool.lp_price();
```

Comment:
💀🚫 The core principle of check balance is that the balance can't be manipulated by an external party. I think that `lp_price()` can be manipulated, since it contains a factor that has to do with the current price in the curve pool, and that can be changed by trades?

Will have to check that `2x virtual_price` still behaves like we need it to behave. At least it should be up only, unlike `lp_price`.

Reply:
that is a good point and I agree. I think that the `lp_price` doesn't only reflect the current state of the pool but has some trailing effect of the previous pool balances.

Changing it to virtual_price * 2 still lets the basic `mint` & `withdrawal` tests to pass.

I think that the purpose of the price oracle is for the pool to be much more susceptible flash loan attacks. Which could be a big security gain for us. Though I don't understand the whole of the Curve Pool's code in detail yet to be able to confidently make that claim.